File: backend/models/Event.js
const mongoose = require('mongoose');

const eventSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Название события обязательно'],
    trim: true,
    maxlength: [200, 'Название не должно превышать 200 символов']
  },
  description: {
    type: String,
    required: [true, 'Описание события обязательно'],
    maxlength: [1000, 'Описание не должно превышать 1000 символов']
  },
  type: {
    type: String,
    enum: ['cleanup', 'workshop', 'conference', 'webinar', 'tree_planting', 'other'],
    required: true
  },
  organizer: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  date: {
    type: Date,
    required: [true, 'Дата события обязательна']
  },
  endDate: {
    type: Date
  },
  location: {
    address: String,
    city: String,
    country: {
      type: String,
      default: 'Россия'
    },
    coordinates: {
      type: [Number], // [longitude, latitude]
      index: '2dsphere'
    }
  },
  isOnline: {
    type: Boolean,
    default: false
  },
  onlineLink: String,
  capacity: {
    type: Number, // максимальное количество участников
    min: 1
  },
  registered: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  attendees: [{ // кто действительно посетил
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  thumbnail: String,
  tags: [String],
  status: {
    type: String,
    enum: ['draft', 'pending', 'approved', 'cancelled'],
    default: 'pending'
  },
  points: {
    type: Number, // награда за участие
    default: 100,
    min: 0
  }
}, {
  timestamps: true
});

// Индексы
eventSchema.index({ date: 1, status: 1 });
eventSchema.index({ organizer: 1 });
eventSchema.index({ type: 1, status: 1 });
eventSchema.index({ 'location.coordinates': '2dsphere' });

// Виртуальное поле для количества зарегистрированных
eventSchema.virtual('registeredCount').get(function() {
  return this.registered.length;
});

// Виртуальное поле для проверки заполненности
eventSchema.virtual('isFull').get(function() {
  if (!this.capacity) return false;
  return this.registered.length >= this.capacity;
});

// Виртуальное поле для проверки прошло ли событие
eventSchema.virtual('isPast').get(function() {
  return new Date() > (this.endDate || this.date);
});

// Метод для регистрации на событие
eventSchema.methods.registerUser = function(userId) {
  // Проверяем, не зарегистрирован ли уже
  if (this.registered.includes(userId)) {
    throw new Error('Вы уже зарегистрированы на это событие');
  }

  // Проверяем вместимость
  if (this.capacity && this.registered.length >= this.capacity) {
    throw new Error('Событие заполнено');
  }

  // Проверяем, не прошло ли событие
  if (this.isPast) {
    throw new Error('Событие уже прошло');
  }

  this.registered.push(userId);
  return this.save();
};

// Метод для отмены регистрации
eventSchema.methods.unregisterUser = function(userId) {
  const index = this.registered.indexOf(userId);
  if (index > -1) {
    this.registered.splice(index, 1);
  }
  return this.save();
};

// Метод для отметки посещения
eventSchema.methods.markAttendance = function(userId) {
  if (!this.registered.includes(userId)) {
    throw new Error('Пользователь не зарегистрирован на событие');
  }

  if (!this.attendees.includes(userId)) {
    this.attendees.push(userId);
  }
  return this.save();
};

module.exports = mongoose.model('Event', eventSchema);

File: frontend/lib/models/event.dart
class Event {
  final String id;
  final String title;
  final String description;
  final String type;
  final String organizerId;
  final String? organizerName;
  final DateTime date;
  final DateTime? endDate;
  final EventLocation? location;
  final bool isOnline;
  final String? onlineLink;
  final int? capacity;
  final List<String> registered;
  final List<String> attendees;
  final String? thumbnail;
  final List<String> tags;
  final String status;
  final int points;
  final DateTime createdAt;

  Event({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.organizerId,
    this.organizerName,
    required this.date,
    this.endDate,
    this.location,
    required this.isOnline,
    this.onlineLink,
    this.capacity,
    required this.registered,
    required this.attendees,
    this.thumbnail,
    required this.tags,
    required this.status,
    required this.points,
    required this.createdAt,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['_id'],
      title: json['title'],
      description: json['description'],
      type: json['type'],
      organizerId: json['organizer'] is String 
          ? json['organizer'] 
          : json['organizer']['_id'],
      organizerName: json['organizer'] is Map 
          ? json['organizer']['username'] 
          : null,
      date: DateTime.parse(json['date']),
      endDate: json['endDate'] != null 
          ? DateTime.parse(json['endDate']) 
          : null,
      location: json['location'] != null 
          ? EventLocation.fromJson(json['location']) 
          : null,
      isOnline: json['isOnline'] ?? false,
      onlineLink: json['onlineLink'],
      capacity: json['capacity'],
      registered: List<String>.from(json['registered'] ?? []),
      attendees: List<String>.from(json['attendees'] ?? []),
      thumbnail: json['thumbnail'],
      tags: List<String>.from(json['tags'] ?? []),
      status: json['status'] ?? 'pending',
      points: json['points'] ?? 0,
      createdAt: DateTime.parse(json['createdAt']),
    );
  }

  int get registeredCount => registered.length;
  bool get isFull => capacity != null && registered.length >= capacity!;
  bool get isPast => DateTime.now().isAfter(endDate ?? date);
  bool get isUpcoming => DateTime.now().isBefore(date);
  bool get isApproved => status == 'approved';
}

class EventLocation {
  final String? address;
  final String? city;
  final String? country;
  final List<double>? coordinates; // [longitude, latitude]

  EventLocation({
    this.address,
    this.city,
    this.country,
    this.coordinates,
  });

  factory EventLocation.fromJson(Map<String, dynamic> json) {
    return EventLocation(
      address: json['address'],
      city: json['city'],
      country: json['country'],
      coordinates: json['coordinates'] != null 
          ? List<double>.from(json['coordinates']) 
          : null,
    );
  }
}

